tests/cases/compiler/contextualSignatureInArrayElementPrefersArrayUnionMemberLibEs2015.ts

```diff
@@ -1,15 +0,0 @@
-// @strict: true
-// @noEmit: true
-// @lib: es2015
-
-// repro from #52588
-
-declare function test(
-  arg: Record<string, (arg: string) => void> | Array<(arg: number) => void>
-): void;
-
-test([
-  (arg) => {
-    arg; // number
-  },
-]);
```

Comment:
I see @ahejlsberg's point about string index signatures contributing to the contextual type of arrays, it's kinda how it works. At the same time, I find it somewhat hard to believe that the user's intention would ever be for this to happen in such a mixed union.  

This doesn't happen to objects (as the `Array` type doesn't contribute to their contextual type here) and it feels wrong that the opposite can't be expressed. We can mix a string and a number signature to achieve this:
```ts
declare function test(
  arg:
    | (Record<string, (arg: string) => void> & { [index: number]: never })
    | Array<(arg: number) => void>
): void;

test([
  (arg) => {
    arg; // number, it works!
  },
]);
```

But that's not quite intuitive and a lot of users wouldn't figure this out on their own as a solution.

It also has a drawback... those become errors:
```ts
test({
  1: (arg) => {}, // error
});

test({
  "2": (arg) => {}, // error
});
```

I feel like if this test case doesn't represent the desired behavior then there is a clear need for something in the language to represent non-arrays. I understand that almost everything is an object... but that's not exactly how people think about those things. Arrays are special enough to not be thought of as plain objects 99% of the time.

Either way... if this is how things should work and if no changes are planned to the current behavior, then I feel like those expectations should be codified in the test suite. So instead of reverting the whole change, I would just revert the `checker.ts` and adjust the tests.

Reply:
We generally only discriminate contextual types based on discriminant properties with explicit literal values in the source type. In an ideal world we'd discriminate a contextual union type based on assignability of the source type to each individial constituent, but given that the contextual type influences the source type, that quickly becomes circular. So, as a rule, we don't do that.

While an object-but-not-array type might be nice to have (object-but-not-function is another one that comes to mind), it is not something we're equipped to consistently reason about. The notion that an array is just an object with numerically named properties is quite deeply engrained in JS/TS.

Anyway, I'm fine with just reverting the changes in `checker.ts` and keeping the tests.